src/Storage/QueryBuilder.php

```diff
@@ -0,0 +1,169 @@
+    $result = $this->database->query($sql, [
+      ':id' => $entity->id(),
+      ':revision_id' => $entity->getRevisionId() ?: $entity->id(),
+    ]);
+    $records = $result->fetchAll(\PDO::FETCH_CLASS, '\Drupal\entity_hierarchy\Storage\Record');
+    $type = $this->fieldStorageDefinition->getTargetEntityTypeId();
+    array_walk($records, function (Record $record) use ($type) {
+      $record->setType($type);
+    });
+    if (!$this->fieldStorageDefinition->isBaseField()) {
+      // Drupal doesn't store an empty row for a NULL parent. Add it back in.
+      if ($first_record = reset($records)) {
+        $record = Record::create($type, $first_record->getTargetId(), 0, $first_record->getDepth() - 1);
+      }
+      else {
+        $record = Record::create($type, $entity->id(), 0, 0);
+      }
+      array_unshift($records, $record);
+    }
+    return new RecordCollection($records);
+  }
+
+  public function findRoot(ContentEntityInterface $entity): ?Record {
+    $ancestors = $this->findAncestors($entity);
+    // Convert FALSE to NULL.
+    return $ancestors->getIterator()->current() ?: NULL;
+  }
+
+  public function findParent(ContentEntityInterface $entity): ?ContentEntityInterface {
+    return $entity->get($this->fieldStorageDefinition->getName())->entity;
+  }
+
+  public function findDepth(ContentEntityInterface $entity): int {
+    $sql = $this->getAncestorSql() . " SELECT depth FROM ancestors ORDER BY depth LIMIT 1";
+    $result = $this->database->query($sql, [
+      ':id' => $entity->id(),
+      ':revision_id' => $entity->getRevisionId() ?: $entity->id(),
+    ]);
+    if ($object = $result->fetchObject()) {
+      $depth = $object->depth * -1;
+      if (!$this->fieldStorageDefinition->isBaseField()) {
+        // Non base fields don't have a record where parent is null. Compensate.
+        $depth = $depth + 1;
+      }
+      return $depth;
+    }
+    return 0;
+  }
+
+  protected function getDescendantSql(): string {
+    $table_name = $this->getTablePrefix() . $this->tables['entity'];
+    $column_id = $this->columns['id'];
+    $column_revision_id = $this->columns['revision_id'];
+    $column_target_id = $this->columns['target_id'];
+    $column_weight = $this->columns['weight'];
+    $sql = <<<CTESQL
+WITH RECURSIVE descendants AS
+(
+  SELECT $column_id as id, $column_revision_id as revision_id, $column_target_id as target_id, $column_weight as weight, 1 AS depth
+  FROM $table_name
+  WHERE $column_target_id = :target_id
+  UNION ALL
+  SELECT c.$column_id, c.$column_revision_id, c.$column_target_id, c.$column_weight, descendants.depth+1
+  FROM $table_name c
+  JOIN descendants ON descendants.id=c.$column_target_id
+)
+CTESQL;
+    return $sql;
+  }
+
+  /**
+   */
+  public function findDescendants(ContentEntityInterface $entity, int $depth = 0, int $start = 1): RecordCollection {
+    $sql = $this->getDescendantSql() . " SELECT id, revision_id, target_id, weight, depth FROM descendants WHERE depth >= :start";
```

Comment:
What happens here if the entity is not the default revision? i.e. a draft - the query only uses the entity table - should it use the revision table for non default revisions?

Reply:
Currently it only checks the default revisions. And I believe this to be accurate because Drupal by default also only checks current revisions (loads, titles, etc) unless you specifically ask for something else. I also believe this is what current EH does (I don't see drafts showing up in DCS side navigation, even as admin role).
Happy to look at a `->findDescendantRevisions()` as a future improvement. Performance will obviously not be as great.